### PR TITLE
Improve worker node provisioning

### PR DESCRIPTION
**Pull request by daxmc99**
These commits help to improve worker node provisioning by reducing unnecessary updates
Addresses https://github.com/rancher/rancher/issues/22669

### CODE REVIEW

File: pkg/controllers/management/nodepool/nodepool.go [hunk 1]
```diff
@@ -319,12 +320,23 @@ func needRoleUpdate(node *v3.Node, nodePool *v3.NodePool) bool {
...
+		nodeRolesMap[services.WorkerRole] = false
+	}
 
 	poolRolesMap := map[string]bool{}
 	poolRolesMap[services.ETCDRole] = nodePool.Spec.Etcd
 	poolRolesMap[services.ControlRole] = nodePool.Spec.ControlPlane
 	poolRolesMap[services.WorkerRole] = nodePool.Spec.Worker
-	return !reflect.DeepEqual(nodeRolesMap, poolRolesMap)
+
+	r := !reflect.DeepEqual(nodeRolesMap, poolRolesMap)
```
Comment (alena1108) [inline pkg/controllers/management/nodepool/nodepool.go]: might be worth printing the discrepancies in the debug mode. It helped with debugging similar issues in the past

Comment (daxmc99) [reply pkg/controllers/management/nodepool/nodepool.go]: What about printing only when it will cause an update?

Comment (alena1108) [reply pkg/controllers/management/nodepool/nodepool.go]: sounds good

File: pkg/controllers/user/nodesyncer/nodessyncer.go [hunk 2]
```diff
@@ -121,7 +121,12 @@ func (n *nodeSyncer) needUpdate(key string, node *corev1.Node) (bool, error) {
 		return true, nil
 	}
 	if existing.Annotations[annotationName] == "" {
+		existing = existing.DeepCopy()
+		if existing.Annotations == nil {
+			existing.Annotations = make(map[string]string)
+		}
 		existing.Annotations[annotationName] = "true"
+		logrus.Info("updating machine with annotation")
```
[the remark this answers is not in the document]
Comment (daxmc99) [reply pkg/controllers/user/nodesyncer/nodessyncer.go]: dropped

File: pkg/controllers/management/nodepool/nodepool.go [hunk 3]
```diff
@@ -292,6 +292,7 @@ func (c *Controller) createOrCheckNodes(nodePool *v3.NodePool, simulate bool) (b
 			changed = true
 			_, err := c.updateNodeRoles(n, nodePool, simulate)
 			if err != nil {
+				logrus.Errorf("error nodepool update %v", err)
```
[the remark this answers is not in the document]
Comment (daxmc99) [reply pkg/controllers/management/nodepool/nodepool.go]: dropped

File: pkg/controllers/management/nodepool/nodepool.go [hunk 4]
```diff
@@ -319,12 +320,26 @@ func needRoleUpdate(node *v3.Node, nodePool *v3.NodePool) bool {
 			nodeRolesMap[services.WorkerRole] = true
 		}
 	}
+	if _, ok := nodeRolesMap[services.ETCDRole]; !ok {
```
[the remark this answers is not in the document]
Comment (daxmc99) [reply pkg/controllers/management/nodepool/nodepool.go]: Thanks, nice catch

**Comment by dramich** [comment]
lgtm

**Pull request merged**